zilencer/management/commands/populate_db.py

```diff
@@ -808,6 +811,46 @@ def send_messages(messages: List[Message]) -> None:
     do_send_messages([{'message': message} for message in messages])
     settings.USING_RABBITMQ = True
 
+def add_emoji_reactions_to_messages(
+        messages: List[Message],
+        users: Optional[Iterable[UserProfile]] = None,
+        emojis: Optional[Iterable[Tuple[str, str]]] = None,
+        reaction_threshold: float = 0.0012) -> None:
+    '''Randomly add emoji reactions to each message from a list.
+
+    For each possible combination of user, message, and emoji, assigns
+    a reaction with that combination if a uniformly distributed random
+    number between 0 and 1 is below the given reaction threshold.
+
+    Note that the reaction threshold will be much smaller than the actual
+    fraction of messages with at least one reaction. See test_populate_db
+    for a test that ensures that the proper fraction of messages have
+    reactions.
+    '''
+    if emojis is None:
+        emojis = [
+            ('+1', '1f44d'),
+            ('smiley', '1f603'),
+            ('eyes', '1f440'),
+            ('crying_cat_face', '1f63f'),
+            ('arrow_up', '2b06'),
+            ('confetti_ball', '1f38a'),
+            ('hundred_points', '1f4af'),
+        ]
+    if users is None:
+        users = UserProfile.objects.filter(is_bot=False)
+    Reaction.objects.bulk_create([
+        Reaction(
+            user_profile=user,
+            message=message,
+            emoji_name=emoji[0],
+            emoji_code=emoji[1],
+            reaction_type=Reaction.UNICODE_EMOJI
+        )
+        for (message, user, emoji) in itertools.product(messages, users, emojis)
```

Comment:
I think this algorithm is going to result in a performance problem if one passes the options to create 1000 users and 100000 messages, which is an important `populate_db` use case.  

We also likely want to limit the users reacting to those who are subscribed to the stream where the message was sent; otherwise we're violating invariants for how Zulip works.

Finally, in an ideal world this algorithm would sometimes generate upvoted reactions (i.e. cases where 2 users added the same reaction to a message), since that gets displayed in the Zulip UI.  Maybe we get that from just having 7 emoji that are used in the list above?

One theory is that we should just have a loop over messages, each one has an x% chance of having at least one reaction, and then add one user and then have a 50% (repeating) chance of repeating the process on that message.

Reply:
Thanks for the suggestions. I'll see if I can use them to update the algorithm.